pkg/kubelet/cm/devicemanager/manager.go

```diff
@@ -907,7 +907,11 @@ func (m *ManagerImpl) allocateContainerResources(pod *v1.Pod, container *v1.Cont
 	}
 
 	// Checkpoints device to container allocation information.
-	return m.writeCheckpoint()
+	if allocatedDevicesUpdated {
```

Comment:
So looking through the code, I don't think this would be the right check. This variable is set to true as soon as you see a device managed by the device Manager, e.g:
```
       allocatedDevicesUpdated  := false
	for k, v := range container.Resources.Limits {
		resource := string(k)
		needed := int(v.Value())
		klog.V(3).Infof("needs %d %s", needed, resource)
		if !m.isDevicePluginResource(resource) {
			continue
		}
		// Updates allocatedDevices to garbage collect any stranded resources
		// before doing the device plugin allocation.
		if !allocatedDevicesUpdated {
			m.UpdateAllocatedDevices()
			allocatedDevicesUpdated = true
		}
```

Reply:
It appear the newest change implements what you suggest in (1) in a slightly different way. The overall logic is the same though.

/approve
/lgtm